src/auth/src/credentials/idtoken/verifier.rs

```diff
@@ -0,0 +1,517 @@
+    /// # use std::time::Duration;
+    /// let audience = "https://example.com";
+    /// let verifier = Builder::new(audience)
+    ///     .with_clock_skew(Duration::from_secs(60))
+    ///     .build();
+    /// ```
+    pub fn with_clock_skew(mut self, clock_skew: Duration) -> Self {
+        self.clock_skew = Some(clock_skew);
+        self
+    }
+
+    /// Verifies the ID token and returns the claims.
+    pub fn build(self) -> Verifier {
+        Verifier {
+            jwk_client: JwkClient::new(),
+            audience: self.audience.clone(),
+            email: self.email.clone(),
+            jwks_url: self.jwks_url.clone(),
+            clock_skew: self.clock_skew.unwrap_or_else(|| Duration::from_secs(10)),
+        }
+    }
+}
+
+/// Verifier is used to verify OIDC ID Tokens.
+///
+/// # Example
+///
+/// ```
+/// # use google_cloud_auth::credentials::idtoken::verifier::Builder;
+/// # use std::time::Duration;
+///
+/// async fn verify_id_token(token: &str) {
+///     let audience = "https://example.com";
+///     let verifier = Builder::new(audience).build();
+///
+///     let claims = verifier.verify(token).await.expect("Failed to verify ID token");
+///     println!("Verified claims: {:?}", claims);
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Verifier {
+    jwk_client: JwkClient,
+    audience: String,
+    email: Option<String>,
+    jwks_url: Option<String>,
+    clock_skew: Duration,
+}
+
+impl Verifier {
+    /// Verifies the ID token and returns the claims.
+    pub async fn verify(&self, token: &str) -> std::result::Result<Map<String, Value>, Error> {
+        let header = jsonwebtoken::decode_header(token).map_err(Error::decode)?;
+
+        let key_id = header.kid.ok_or_else(|| Error::missing_field("kid"))?;
+
+        let mut validation = Validation::new(header.alg);
+        validation.leeway = self.clock_skew.as_secs();
+        validation.set_issuer(&["https://accounts.google.com", "accounts.google.com"]);
+        validation.set_audience(std::slice::from_ref(&self.audience));
+
+        let expected_email = self.email.clone();
+        let jwks_url = self.jwks_url.clone();
+
+        let cert = self
+            .jwk_client
+            .get_or_load_cert(key_id, header.alg, jwks_url)
+            .await
+            .map_err(Error::load_cert)?;
+
+        let token = jsonwebtoken::decode::<Map<String, Value>>(&token, &cert, &validation)
+            .map_err(Error::invalid)?;
+
+        let claims = token.claims;
+        if let Some(email) = expected_email {
+            let email_verified = claims["email_verified"].as_bool().unwrap_or(false);
+            if !email_verified {
+                return Err(Error::invalid("email not verified"));
+            }
+            let token_email = claims["email"]
+                .as_str()
+                .ok_or_else(|| Error::missing_field("email"))?;
+            if !email.eq(token_email) {
+                return Err(Error::invalid("invalid email"));
```

Comment:
Hmmm... not awesome, should the variant for `Error::invalid()` capture what is invalid in a more specific type?